### PR TITLE
libavutil/hwcontext_qsv: Add width and height alignment

### PR DESCRIPTION
**Pull request by wenbinc-Bin**
Use cropping to store real w and h. Use width and height to store
allocated buffer size.

### CODE REVIEW

**Comment by wenbinc-Bin** [comment]
Hi @xhaihao. I align width and height when initializing hw_frames like this. This can fix the hwupload -> hwdownload regression and no need to call get_buffer().

**Comment by xhaihao** [comment]
After applying this patch, the crop info is wrong 

In addition, this fix is not needed for qsv hwupload, the command below may works without this patch

```
ffmpeg -f lavfi -init_hw_device qsv=qsv -i yuvtestsrc=size=1920x1080 -vf "format=nv12,hwupload=extra_hw_frames=16" -f null -
```
I presumed there is a hidden issue in qsv hwdownload.

**Comment by wenbinc-Bin** [comment]
@xhaihao I updated the commit. How about I fix issue like this?

File: libavutil/hwcontext_qsv.c [hunk 1]
```diff
@@ -1063,6 +1063,34 @@ static int qsv_transfer_data_from(AVHWFramesContext *ctx, AVFrame *dst,
...
+     * Height must be a multiple of 16 for progressive frame sequence and a
+     * multiple of 32 otherwise.", so allign all frames to 16 before downloading. */
+    if (ctx->height & 15 || dst->linesize[0] & 15) {
+        AVFrame *tmp_frame;
+        tmp_frame = av_frame_alloc();
+        if (!tmp_frame)
+            return AVERROR(ENOMEM);
+        ret = av_frame_ref(tmp_frame, dst);
+        if (ret < 0)
+            return ret;
```
Comment (xhaihao) [inline libavutil/hwcontext_qsv.c:1077]: call `av_frame_free(&tmp_frame)` before returning an error.

File: libavutil/hwcontext_qsv.c [hunk 2]
```diff
@@ -1063,6 +1063,34 @@ static int qsv_transfer_data_from(AVHWFramesContext *ctx, AVFrame *dst,
...
+        av_frame_unref(dst);
+
+        dst->width  = FFALIGN(tmp_frame->width, 16);
+        dst->height = FFALIGN(ctx->height, 16);
+        dst->format = tmp_frame->format;
+        ret = av_frame_get_buffer(dst, 0);
+        if (ret < 0)
+            return ret;
```
Comment (xhaihao) [inline libavutil/hwcontext_qsv.c:1087]: call `av_frame_free(&tmp_frame)` too when returning an error.

File: libavutil/hwcontext_qsv.c [hunk 3]
```diff
@@ -1063,6 +1063,34 @@ static int qsv_transfer_data_from(AVHWFramesContext *ctx, AVFrame *dst,
...
+
+        dst->width = tmp_frame->width;
+        dst->height = tmp_frame->height;
+        ret = av_frame_copy_props(dst, tmp_frame);
+        if (ret < 0)
+            return ret;
```
Comment (xhaihao) [inline libavutil/hwcontext_qsv.c:1095]: call `av_frame_free(&tmp_frame)` too when returning an error.

File: libavutil/hwcontext_qsv.c (hunk 1 above)
Comment (wenbinc-Bin) [reply libavutil/hwcontext_qsv.c:1077]: Done. Thanks

File: libavutil/hwcontext_qsv.c (hunk 2 above)
Comment (wenbinc-Bin) [reply libavutil/hwcontext_qsv.c:1087]: Done. Thanks

File: libavutil/hwcontext_qsv.c (hunk 3 above)
Comment (wenbinc-Bin) [reply libavutil/hwcontext_qsv.c:1095]: Done. Thanks

File: libavutil/hwcontext_qsv.c (hunk 1 above)
Comment (xhaihao) [reply libavutil/hwcontext_qsv.c:1077]: I am sorry calling `av_frame_free(&tmp_frame)` here is not needed here.

**Review by xhaihao — APPROVED** [review]
LGTM

**Comment by xhaihao** [comment]
@wenbinc-Bin please check why pre-check failed

**Comment by wenbinc-Bin** [comment]
merged https://patchwork.ffmpeg.org/project/ffmpeg/patch/20220406101008.1128823-1-wenbin.chen@intel.com/

**Pull request closed without merging**